Add FavouriteCities search controller

vandermesis opened the pull request.

vandermesis committed:
Add SearchControllerHelper

vandermesis committed:
Inject SearchControllerHelper to FavouriteCitiesController

vandermesis committed:
Setup SearchController for FavouriteCitiesController

vandermesis committed:
Add cities search methods for FavouriteCitiesInteractor and FavouriteCitiesPresenter

vandermesis committed:
Fix all cities not showing when user delete search text

vandermesis committed:
Fix all cities displayed after selecting cell

vandermesis committed:
Fix FavouriteCitiesInteractorTests

vandermesis committed:
Add FavouriteCitiesInteractorTests unit tests for search method

vandermesis committed:
Add FavouriteCitiesPresenterTests unit tests for search method

vandermesis committed:
Add Scope Bar to filter favourite cities

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesController.swift:
```diff
@@ -66,7 +77,29 @@ extension FavouriteCitiesController: UITableViewDataSource {
+extension FavouriteCitiesController: UISearchBarDelegate {
+
+    func searchBar(_ searchBar: UISearchBar, selectedScopeButtonIndexDidChange selectedScope: Int) {
+        var state: Bool {
```
state? state of what, rename

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesPresenter.swift:
```diff
@@ -9,17 +9,35 @@
-    func presentCities(allCities: [City], favourites: [City]) {
+    func presentCities(allCities: [City], favourites: [City], cityName: String, filteringState: Bool) {
+        let mergedCities = mergeFavouriteCities(allCities: allCities, favourites: favourites)
+        let searchedCities = mergedCities.filter { (cities: FavouriteCitiesListDisplayable) -> Bool in
```
you're calculating these but you dont need to do it when filteringState is set/not set, do not do things you don't need at the certain moment

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesPresenter.swift:
```diff
@@ -9,17 +9,35 @@
 extension FavouriteCitiesPresenterImpl: FavouriteCitiesPresenter {
 
-    func presentCities(allCities: [City], favourites: [City]) {
+    func presentCities(allCities: [City], favourites: [City], cityName: String, filteringState: Bool) {
```
why not using 1 argument:
filteringPhrase: String?
which also determines filering state? if nil = you know that it's false
use 1 parameter, not 2 everywhere

SredzinskiPatryk commented on CleanWeather/Support/en.lproj/Localizable.strings:
```diff
@@ -24,7 +24,12 @@
 "now" = "Now";
 "favouriteCities" = "Favourite Cities";
+"all" = "All";
+"favourites" = "Favourites";
```
you could think about doing these localizable strings in a nicer way - but in a separate PR

I am used to:

"FavouriteCities.Title" = "Favourite Cities";
"FavouriteCities.Favourites = "Favourites";
"FavouriteCities.All = "All";
"FavouriteCities.SearchPlaceholder = "Search Cities";

etc

vandermesis committed:
Fix FavouriteCitiesPresenter presentCities did calculations when not needed

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

vandermesis committed:
Fix FavouriteCities scene scope bar to filter favourite cities

vandermesis committed:
Refactor FavouriteCities scene filtering methods with less parameters

vandermesis committed:
Add FavouriteCitiesPresenter unit tests for filtering functionality

vandermesis committed:
Add TableViewControllerExtension method to animate table view reloading

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesPresenter.swift:
```diff
@@ -9,17 +9,57 @@
+            if favouriteState == true {
+                let filteredCities = mergedCities.filter { $0.isFavourite }
+                controller?.displayCities(filteredCities)
+            } else {
```
split this method to a few shorter ones

SredzinskiPatryk commented on CleanWeatherTests/Scenes/FavouriteCities/FakeFavouriteCitiesPresenter.swift:
```diff
@@ -15,10 +15,14 @@ final class FakeFavouriteCitiesPresenter: FakePresenter, FavouriteCitiesPresente
+    var presentCitiesFavouriteStateCalled: Bool?
 
-    func presentCities(allCities: [City], favourites: [City]) {
+    func presentCities(allCities: [City], favourites: [City], filteringPhrase: String?, favouriteState: Bool?) {
```
how about making it a bit better with custom structure?

vandermesis committed:
Add FavouriteCities model to refactor FavouriteCities scene presentCities method

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Conflicts:
	CleanWeather.xcodeproj/project.pbxproj
	CleanWeather/Networking/Base/APIClient.swift
	CleanWeather/Networking/WeatherNetworking.swift

vandermesis committed:
Modify search controller setup

Modified:
	SearchBar and ScopeBar always visible
	ScopeBar is set to all cities when viewDidLoad

vandermesis committed:
Refactor FavouriteCitiesPresenter filtering method

Split to methods:
	filerCities
	searchCitiesForPhrase
	searchCitiesForFavouriteState

SredzinskiPatryk commented on CleanWeather/Models/FavouriteCities.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+struct FavouriteCities {
```
I didn't mean to build a structure from all the parameters
I just wantd you to have a filteringPhrasee and favouriteState separated named maybe: CitiesFilter with string phrase and favouritesSelected: Bool
Current structure is weird, I've no idea what is it for, and has converter to Displayable model which is always connected to single scene, structure like that should be reusable

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesController.swift:
```diff
@@ -17,12 +17,17 @@ final class FavouriteCitiesController: SharedViewController {
-
     private var citiesDataSource = [FavouriteCitiesListDisplayable]()
+    private var isFavourite: Bool?
+    private var filteringPhrase: String?
```
I did want that structure to be here, replacing these 2 parameters

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesController.swift:
```diff
@@ -66,7 +72,26 @@ extension FavouriteCitiesController: UITableViewDataSource {
-        interactor.didSelectCity(id: citiesDataSource[indexPath.row].id)
+        interactor.didSelectCity(id: citiesDataSource[indexPath.row].id,
+                                 filteringPhrase: filteringPhrase,
+                                 favouriteState: isFavourite)
```
I did want that structure to be here, replacing these 2 parameters

SredzinskiPatryk commented on CleanWeather/Scenes/FavouriteCities/FavouriteCitiesController.swift:
```diff
@@ -66,7 +72,26 @@ extension FavouriteCitiesController: UITableViewDataSource {
+        let filteringPhrase = searchController.searchBar.text
+        self.filteringPhrase = filteringPhrase
+        interactor.filterFavouriteCities(filteringPhrase: filteringPhrase,
+                                         favouriteState: isFavourite)
```
I did want that structure to be here, replacing these 2 parameters

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Conflicts:
	CleanWeather/Support/en.lproj/Localizable.strings

vandermesis committed:
Resolve conflicts after merge from remote/master

Conflicts:
	Few keys from localizable.strings where missing

vandermesis committed:
Refactor FavouriteCities scene FavouriteCites model struct

Refactored:
	Change name to CitiesFilter
	Remove allCities and favourites variables
	Move mergedCitiesData computed property to presenter method
	Refactor scene methods to support CitiesFilter model

vandermesis committed:
Fix FavouritesCities scene unit tests after refactoring

vandermesis committed:
Change CitiesFilter initialization strategy

Changes:
	Initialize CitiesFilter in FavouriteCitiesCreator
	Inject CitiesFilter through Constructor Injection using DI
